
---
title: "switchde: inference of switch-like gene behaviour along single-cell trajectories"
author: "Kieran Campbell"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document 
references:
- id: trapnell
  title: The dynamics and regulators of cell fate decisions are revealed by pseudotemporal ordering of single cells.
  author:
  - family: Trapnell
    given: Cole
  container-title:  Nature biotechnology
  publisher: Nature Publishing Group
  type: article-journal
  issued:
    year: 2014
---

```{r load-data, cache = FALSE, message = FALSE, warning = FALSE, include = FALSE}
library(dplyr)
library(monocle)
knitr::opts_chunk$set( cache = TRUE )
devtools::load_all("~/oxford/switch/switchde/") ## REMOVE WHEN COMPLETED
```

  
# Introduction

`switchde` is an `R` package for detecting switch-like differential expression along single-cell RNA-seq trajectories. It assumes genes follow a sigmoidal pattern of gene expression and tests for differential expression using a likelihood ratio test. It also returns maximum likelihood estimates (MLE) for the sigmoid parameters, which allows filtering of genes for up or down regulation as well as where along the trajectory the regulation occurs.

The parametric form of gene expression assumed is a sigmoid:

```{r sigmoid-plot, fig.width = 4, fig.height = 3, warning = FALSE}
example_sigmoid()
```

Governed by three parameters:

* $\mu_0$ The half-peak expression
* $k$ The 'activation strength'. If positive, the gene is upregulated along the trajectory; if negative, the gene is downregulated. The magnitude of $k$ corresponds to how fast the gene is up or down regulated.
* $t_0$ The 'activation time', or where in the trajectory this behaviour occurs. Note this parameter should be interpreted with respect to the overall range of the pseudotimes supplied.



# Example on real dataset

We provide a brief example on the single-cell RNA-seq dataset of differentiating myoblasts from  [@trapnell]. This data is stored in `HSMM` in the Bioconductor package `HSMMSingleCell`. 



We first subset down to only genes expressed in at least 10% of cells, then from that set down to a further 5000 for computational convenience. We then transform the data to `log2(FPKM + 1)`.

```{r subset-genes}
set.seed(123L)
library(HSMMSingleCell)
data(HSMM)
HSMM <- HSMM[rowMeans(exprs(HSMM) > 0) > 0.1, ] # gene expressed in 10% of cells
HSMM <- HSMM[sample(seq_len(nrow(HSMM)), 1000), ] # sample down to 5000 for computational convenience

x <- log2(exprs(HSMM) + 1)
pseudotime <- HSMM$Pseudotime
```

Model fitting and differential expression testing is provided by a call to the `switchde` function.

```{r test-de}
sde <- switchde(x, pseudotime)
```

This returns a `data.frame` with 6 columns:

* `gene` The gene name, taken from either `featureNames(sce)` or `rowNames(X)`
* `pval` The p-value associated with differential expression
* `qval` The Benjamini-Hochberg corrected q-value associated with differential expression
* `mu0` The MLE estimate of $\mu_0$
* `k` The MLE estimate of $k$
* `t0` The MLE estimate of $t_0$

We can use the function `arrange` from `dplyr` to order this by q-value:

```{r view-results}
arrange(sde, qval)
```

We may then wish to plot the expression of a particular gene and the MLE model. This is acheived using the `switchplot` function, which takes three arguments:

* `x` Vector of expression values
* `pseudotime` Pseudotime vector of the same length as `x`
* `pars` The (`mu_0`, `k`, `t0`) parameter tuple

We can easily extract the parameters using the `extract_pars` function and pass this to `switchplot`:

```{r plot, fig.width = 5, fig.height = 3}
gene <- sde$gene[which.min(sde$qval)]
pars <- extract_pars(sde, gene)
print(pars)

switchplot(x[gene, ], HSMM$Pseudotime, pars)
```

Note that `switchplot` returns a `ggplot` which can be further customised (e.g. using `theme_xxx()`, etc).


# Technical info

```{r session-info}
sessionInfo()
```

# References

